src/main/java/org/reaktivity/nukleus/amqp/internal/stream/AmqpServerFactory.java

```diff
@@ -732,15 +783,55 @@ private int decodePlainFrame(
                 break decode;
             }
 
-            final AmqpPerformativeFW performative = frameHeader.performative();
-            final AmqpDescribedType descriptor = performative.kind();
-            final AmqpServerDecoder decoder = decodersByPerformative.getOrDefault(descriptor, decodeUnknownType);
+            server.decodableBodyBytes = (int) (frameSize - frameHeader.doff() * 4);
             server.decodeChannel = frameHeader.channel();
-            server.decodableBodyBytes = frameSize - frameHeader.doff() * 4;
-            server.decoder = decoder;
+            server.decoder = decodePerformative;
             server.readIdleTimeout = defaultIdleTimeout;
             server.doSignalReadIdleTimeoutIfNecessary();
-            progress = performative.offset();
+            progress = frameHeader.limit();
+        }
+
+        return progress;
+    }
+
+    private int decodePerformative(
+        AmqpServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final int length = limit - offset;
+
+        int progress = offset;
+
+        decode:
+        if (length != 0)
+        {
+            final AmqpPerformativeTypeFW performativeType = amqpPerformativeTypeRO.tryWrap(buffer, offset, limit);
+            if (performativeType == null)
+            {
+                break decode;
+            }
+
+            final AmqpPerformativeType descriptor = performativeType.get();
+
+            final AmqpServer.AmqpSession session = server.sessions.get(server.decodeChannel);
+            if (session != null && session.sessionState == AmqpSessionState.DISCARDING && descriptor != END)
+            {
+                server.decoder = decodeIgnoreFrameBody;
+                break decode;
+            }
+
+            server.decoder = decodersByPerformativeType.getOrDefault(descriptor, decodeUnknownType);
+            server.decodableBodyBytes -= PERFORMATIVE_SIZE;
```

Comment:
Need to add an `assert server.decodableBytes >= 0;`
Also might be better to use `server.decodableBytes -= performativeType.sizeof()`, for example what if it is sent using an 8-byte `ulong` instead of a 1-byte `ulong`?